### PR TITLE
Reload Button

### PR DESCRIPTION
**Pull request by QmelZ**
adds a reload button at the bottom of mod menu that reloads the game.
image

(i made this with very little java knowledge so its probably bad, but i tested it and it works)

### CODE REVIEW

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 import mindustry.ui.*;
 
 import java.io.*;
+import java.net.URISyntaxException;
```
Comment (LottieVixen) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:26]: why is this added?

**Comment by Quezler** [comment]
did you copy the code from the BeControl.java?

if yes, then i didn't recognize it, but not having duplicate code might be cleaner

if no, take a peek there at how its done

**Review by Quezler — CHANGES_REQUESTED** [review]
probably won't work on mac as it is currently

File: core/src/mindustry/ui/dialogs/ModsDialog.java (hunk 1 above)
Comment (QmelZ) [reply core/src/mindustry/ui/dialogs/ModsDialog.java:26]: because of the catching URISyntaxException (i dont really know how it works)

**Comment by QmelZ** [comment]
> did you copy the code from the BeControl.java?

no, i did not actually. if i remembered it existed i would have copied from there

**Comment by QmelZ** [comment]
> probably won't work on mac as it is currently

i dont really have an idea how it works except "it gets the file path and runs `java -jar filepath`"

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 2]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
         addCloseButton();
 
         buttons.button("@mods.guide", Icon.link, () -> Core.app.openURI(modGuideURL)).size(210, 64f);
+        
+        if(Vars.mobile = false) {
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:40]: maybe you mean `==` here?

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 3]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
...
+            buttons.button("@mods.reload", Icon.refresh, () -> {
+            
+                try {
+            
+                String jarPath = getClass().getProtectionDomain().getCodeSource().getLocation().toURI().getPath().replaceFirst("/", "").replaceAll("/", "\\" + File.separator).toString();
+            
+                Runtime rt = Runtime.getRuntime();
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:47]: unnecessary variable, use `Runtime.getRuntime().exec(...)`

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 4]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
...
+                String[] commandAndArguments = {"java","-jar", jarPath};
+                Process p = rt.exec(commandAndArguments);
+            
+                } catch (IOException i) {
+                    i.printStackTrace();
+                } catch (URISyntaxException u) {
+                    u.printStackTrace();
+                };
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:55]: handle any error (i.e `Throwable`), then you don't need to import `URISyntaxException`.

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 5]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
         addCloseButton();
 
         buttons.button("@mods.guide", Icon.link, () -> Core.app.openURI(modGuideURL)).size(210, 64f);
+        
+        if(Vars.mobile = false) {
+            buttons.button("@mods.reload", Icon.refresh, () -> {
+            
+                try {
+            
+                String jarPath = getClass().getProtectionDomain().getCodeSource().getLocation().toURI().getPath().replaceFirst("/", "").replaceAll("/", "\\" + File.separator).toString();
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:45]: indentation. it should be:
```
try {
    // code
}
```

**Comment by Quezler** [comment]
mac needs extra flags (see becontrol)

i would probably suggest to rewrite it to make restart a public void function on becontrol, and just call that via the static vars for your reload button

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 6]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
...
+            
+                Runtime rt = Runtime.getRuntime();
+                String[] commandAndArguments = {"java","-jar", jarPath};
+                Process p = rt.exec(commandAndArguments);
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:49]: unused `p` variable.

**Comment by Quezler** [comment]
though that code is intended for desktop standalone, no idea how it'll behave for steam / ios / android*

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 7]
```diff
@@ -35,6 +36,27 @@ public ModsDialog(){
...
+        
+        if(Vars.mobile = false) {
+            buttons.button("@mods.reload", Icon.refresh, () -> {
+            
+                try {
+            
+                String jarPath = getClass().getProtectionDomain().getCodeSource().getLocation().toURI().getPath().replaceFirst("/", "").replaceAll("/", "\\" + File.separator).toString();
+            
+                Runtime rt = Runtime.getRuntime();
+                String[] commandAndArguments = {"java","-jar", jarPath};
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:48]: why create new variable instead of passing directly to `exec()`?

**Review by summetdev — CHANGES_REQUESTED** [review]

**Comment by QmelZ** [comment]
at least someone told me what's wrong with my bad code, will fix asap

also quezler i will check the thing about macs too

**Comment by Anuken** [comment]
This implementation:
- Doesn't preserve VM arguments, e.g. `-XstartOnFirstThread`
- Doesn't preserve game command-line arguments
- Doesn't work on Android/iOS
- Doesn't preserve the Java version that launched it
- Doesn't work on systems where Java is not installed (which is quite a few machines)
- Doesn't follow the formatting guidelines
- Is unlikely to work properly, as most of these problems are nearly impossible to fix

**Comment by QmelZ** [comment]
fair enough i guess
